contracts-periphery/src/UmbraBatchSend.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+interface IUmbra {
+  function sendEth(
+    address payable receiver,
+    uint256 tollCommitment,
+    bytes32 pkx,
+    bytes32 ciphertext
+  ) external payable;
+
+  function sendToken(
+    address receiver,
+    address tokenAddr,
+    uint256 amount,
+    bytes32 pkx,
+    bytes32 ciphertext
+  ) external payable;
+}
+
+contract UmbraBatchSend {
+  IUmbra internal constant umbra = IUmbra(0xFb2dc580Eed955B528407b4d36FfaFe3da685401);
+
+  struct SendEth {
+    address payable receiver;
+    uint256 amount;
+    bytes32 pkx;
+    bytes32 ciphertext;
+  }
+
+  struct SendToken {
+    address receiver;
+    address tokenAddr;
+    uint256 amount;
+    bytes32 pkx;
+    bytes32 ciphertext;
+  }
+  
+  error ValueMismatch();
+  event Log(address indexed caller, uint256 indexed value, string message);
```

Comment:
A few things:

* Would like to think of a better name for this event, maybe `BatchedSend`
* Nit: `caller` -> `sender` is more descriptive
* I think we need to include the token address in this event
* Nit: `value` -> `amount` is more generic
* What's the string `message` about?

Reply:
I was on the fence at whether or not we should even have a log here at all, but I think it will be useful for UI/UX helpers. I think the use case is to find batch sends from a user. Given that, this doesn't need many details except for the sender:
- Have a single event: `BatchSendExecuted(address indexed caller)`
- Then we search for all those events by the user
- That gives a tx hash/block, then we query the node for `Announcement` events from the same sender in that block

Having this be more complex means you need to emit it each iteration of the loop, which seems like a waste of gas since the `Announcement` contains all the same info. It does reduce an RPC call on the frontend, but that's not worth the gas IMO, especially cause we can write a subgraph to fetch batch-sends and just make one request

Reply:
>  I think the use case is to find batch sends from a user

Well, more specifically, we show the from address on the Receive page. Are you accounting for how we'll do that? Since the msg.sender from the `Umbra` contract's perspective will be the batch send contract, we'll need a way to find it each time.

Given that, I'm actually wondering if we should be emitting all the params of the Umbra announcement each time? Then on transaction scanning we could:

* Fetch all normal `Umbra` Announcement events
* Throw out the ones where sender is the Batch Send contract
* Fetch all Batch Send Announcements
* Combine the remaining Umbra Announcements w/ the Batch Send Announcements

If we don't do it this way, we'll need to do something like what you described above each time to reconcile the sender. It could get messy. Agree the gas overhead is not ideal though.

> That gives a tx hash/block, then we query the node for Announcement events from the same sender in that block

But you could have more than one per block (multiple txs) or even more than one per txn (if a contract batches our Batch Send method 😅 ), right?

Reply:
> Well, more specifically, we show the from address on the Receive page. Are you accounting for how we'll do that? Since the msg.sender from the `Umbra` contract's perspective will be the batch send contract, we'll need a way to find it each time.

`Annoucement` events do not emit a sender parameter, so right now we use the transaction's `from` address as the sender (i.e. `tx.origin`). So that approach wouldn't change here. (Obviously this doesn't work if sends are from a multsig/contract wallet, but that's a separate issue)

So I think my proposal is the simplest/cheapest and accomplishes what we need?

Reply:
Ahh ok, you're right! I think the only edge case to consider then is a batch of batch sends. Any integration logic assuming one `BatchSendExecuted` per txn could break on that, right? But yeah shouldn't be a big issue I guess.

Reply:
That would be a weird edge case, i.e. why would someone do that, but with our logic of "tx.origin = sender" I think everything should still work out fine in that situation?